File: 백준/Gold/1238. 파티/파티.py
import sys
input = sys.stdin.readline
import heapq

def dijkstra(start):
    distance = [1e9] * (N + 1)
    pq = []
    heapq.heappush(pq, (0, start))
    distance[start] = 0
    
    while pq:
        dist, now = heapq.heappop(pq)
        
        if distance[now] >= dist:
            for node, cost in graph[now]:
                new_cost = dist + cost
                
                if distance[node] > new_cost:
                    distance[node] = new_cost
                    heapq.heappush(pq, (new_cost, node))
    if start != X:
            return distance[X]
    return distance


N, M, X = map(int, input().split())
graph = [[] for _ in range(N + 1)]
for _ in range(M):
    f, t, w = map(int, input().split())
    graph[f].append([t, w])


node_to_x = [0] * (N + 1)
for i in range(1, N + 1):
    if i != X:
        node_to_x[i] = dijkstra(i)

x_to_node = dijkstra(X)

result = []
for i in range(1, N + 1):
    if i != X:
        result.append(x_to_node[i] + node_to_x[i])
        
print(max(result))

File: 백준/Gold/1238. 파티/파티.java
import java.io.*;
import java.util.*;

class Node implements Comparable<Node> {
    int idx;
    int cost;

    @Override
    public String toString() {
        return "Node{" +
                "idx=" + idx +
                ", cost=" + cost +
                '}';
    }

    Node(int idx, int cost) {
        this.idx = idx;
        this.cost = cost;
    }

    @Override
    public int compareTo(Node o) {
        return this.cost - o.cost;
    }
}

public class Main {
    private static int N, M, X;
    private static List<List<Node>> arr;

    private static PriorityQueue<Node> pq;

    private static int[] distance;
    public static void main(String[] args) throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(input.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        arr = new ArrayList<>();

        for (int i = 0; i < N + 1; i++) {
            List<Node> lst = new ArrayList<>();
            arr.add(lst);
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(input.readLine());

            int f, t, dist;

            f = Integer.parseInt(st.nextToken());
            t = Integer.parseInt(st.nextToken());
            dist = Integer.parseInt(st.nextToken());

            arr.get(f).add(new Node(t, dist));
        }

        int ans = 0;

        for (int i = 1; i < N + 1; i++) {
            ans = Math.max(ans, dijkstra(i, X) + dijkstra(X, i));
        }

        System.out.println(ans);
    }

    static int dijkstra(int start, int end) {
        distance = new int[N + 1];
        Arrays.fill(distance, Integer.MAX_VALUE);

        pq = new PriorityQueue<>();

        pq.add(new Node(start, 0));

        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node now = pq.poll();

            int node, cost;
            node = now.idx;
            cost = now.cost;

            for (Node next : arr.get(node)) {
                int nextNode, dist;
                nextNode = next.idx;
                dist = next.cost;

                int newCost = cost + dist;

                if (newCost < distance[nextNode]) {
                    distance[nextNode] = newCost;
                    pq.add(new Node(nextNode, newCost));
                }
            }
        }

        return distance[end];
    }
}